### ISSUE TITLE
Accesing network performance

### ISSUE THREAD

**Issue by neoyagami** [mafredri/cdp#44]
Hi.
 does this implementacion have ways to capture all the network events including the performance of each resource loaded?

**Comment by mafredri** [comment]
Hi @neoyagami, this package makes the whole Chrome DevTools Protocol available in Go, so if the protocol support it, you can most definitely do it. I'm not quite sure what you mean with *performance of each resource loaded* but you can definitely capture all network events. E.g. by monitoring the `RequestWillBeSent/ResponseReceived/LoadingFinished/LoadingFailed` events in the `Network` domain. Or you can choose to intercept / modify them with `SetRequestInterception/RequestIntercepted/ContinueInterceptedRequest`, etc.

**Comment by orian** [comment]
I wonder, how likely is that some of the events don't get into the lib? 

I've setup a listener to LoadingFinished but it never gets events for some is the requests. 
It looks like the request went awry.

I leave the listening with 20 seconds timeout. Resources are static and requested early during page handling.

**Comment by mafredri** [comment]
@orian have you called `Network.Enable`?

> I wonder, how likely is that some of the events don't get into the lib?

I'd say it's very unlikely. You can enable logging (see the logging example in GoDoc) to know for sure.

If `Network.Enable` doesn't work, could you open up a new issue with a small piece of code that reproduces it?

**Comment by orian** [comment]
@mafredri yes, I've called Network.Enable. In fact, among about 100 requests I don't get LoadingFinished for ~20 requests. From which some are because of a server dropping connection (mainly Facebook).

So in general `Network.Enable` works.

I'm wondering if this may be related to buffering? Maybe Chrome cannot push events fast enough and the buffer is overfilled, so it drops them?

**Comment by mafredri** [comment]
@orian alright, but you won't receive a `LoadingFinished` event for failed requests, you should check for `LoadingFailed`.

> I'm wondering if this may be related to buffering? Maybe Chrome cannot push events fast enough and the buffer is overfilled, so it drops them?

I don't think that's the case, if Chrome did something like that, the Chrome DevTools (UI) would have trouble working properly. They rely on the state provided by the events. And even in `cdp`, the event buffering uses a kind of unbounded channel, so no events will be dropped (as it can only become full when you run out of memory).

**Comment by orian** [comment]
@mafredri Yes, I know. Already doing it. But the issue I'm describing is about the resources which **loads correctly** (I can tell this from screenshot). 

Before creating a separate issue I'll investigate further, replicate and ensure it's not a bug in my code.

I'll be back soon.

**Comment by mafredri** [comment]
Oh, that sounds worrisome @orian. I'm definitely interested in getting to the bottom of it if this is the case. Wondering if this could be the result of cached content not triggering `LoadingFinished` or similar. You could try listening for `RequestServedFromCache` to see if that helps.

> PS. I like the way you've designed a library and code generator. I had to get used to it, but it's well thought.

Thanks, I'm happy you like it 😄!

**Comment by orian** [comment]
I'm listening to all Network events and saving them in a wrapping structure in `sync.Map` using `RequestID` as a key. Then, later on I'm joining all data I know about the event to tell how the request ended and, in fact, i'm not getting any further events for some of the requests.

I've also disabled all caching through an API.

**Comment by orian** [comment]
@mafredri I've found a bug in my code. Indeed, the requests get cancelled and I'm getting:
```
"LoadingFailed": {
  "requestId": "30742.16",
  "timestamp": 16062.482595,
  "type": "Image",
  "errorText": "net::ERR_ABORTED",
  "canceled": true
},
```
I have a hard time working on it. First, it looks non-deterministic. Second, it seems to be related:

**Comment by mafredri** [comment]
@orian I'm glad you fount the issue.

> May it be some kind prefetching?

Hard to say. I think it's more likely that something is triggering a reload or re-navigation, perhaps from JavaScript.

You could check the `Reason` attribute of `FrameScheduledNavigation` (Page-domain) and see if that provides any clues.
